src/doc/algorithm/shift_image.cpp

```diff
@@ -20,31 +23,101 @@
 namespace doc {
 namespace algorithm {
 
-void shift_image_with_mask(Image* image, const Mask* mask, int dx, int dy)
+void shift_image(Image* image, int dx, int dy, double angle)
 {
-  gfx::Rect bounds = mask->bounds();
-  if (bounds.isEmpty())
-    return;
-
+  gfx::Rect bounds(image->bounds());
+  if (cos(angle) < -sqrt(2)/2) {
+    dx = -dx;
+    dy = -dy;
+  }
+  else if (sin(angle) >= sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = -dy;
+    dy = aux;
+  }
+  else if (sin(angle) < -sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = dy;
+    dy = -aux;
+  }
   // To simplify the algorithm we use a copy of the original image, we
   // could avoid this copy swapping rows and columns.
   ImageRef crop(crop_image(image, bounds.x, bounds.y, bounds.w, bounds.h,
-                           image->maskColor()));
-
-  int u = dx;
-  int v = dy;
-  while (u < 0) u += bounds.w;
-  while (v < 0) v += bounds.h;
+                         image->maskColor()));
 
   for (int y=0; y<bounds.h; ++y) {
     for (int x=0; x<bounds.w; ++x) {
-      put_pixel(image,
-                bounds.x + ((u + x) % bounds.w),
-                bounds.y + ((v + y) % bounds.h),
-                get_pixel(crop.get(), x, y));
+        put_pixel(image,
+                  (bounds.w + dx + x) % bounds.w,
+                  (bounds.h + dy + y) % bounds.h,
+                  get_pixel(crop.get(), x, y));
     }
   }
 }
 
+void shift_image_with_mask(Cel* cel, const Mask* mask, int dx, int dy)
+{
+  ASSERT(!cel->bounds().isEmpty());
+  ASSERT(!mask->bounds().isEmpty());
+
+  // Making a image which bounds are equal to the UNION of cel->bounds() and mask->bounds()
+  gfx::Rect compCelBounds = cel->bounds() | mask->bounds();
+  ImageRef compImage(Image::create(cel->image()->pixelFormat(), compCelBounds.w, compCelBounds.h));
+
+  // Making a Rect which represents the mask bounds of the original MASK relative to
+  // the new COMPOUND IMAGE (compImage)
+  gfx::Point maskCelGap(0, 0);
+  if (cel->bounds().x < mask->bounds().x)
+    maskCelGap.x = mask->bounds().x - cel->bounds().x;
+  else
+    maskCelGap.x = 0;
```

Comment:
We can remove the `else` case because `maskCelGap.x` is already initialized `= 0`.